pallets/governance-origins/src/lib.rs

```diff
@@ -0,0 +1,191 @@
+		/// Maximum amount able to be spent by SmallSpender origin from treasury at once
+		#[pallet::constant]
+		type MaxSmallSpenderSpend: Get<BalanceOf<Self>>;
+		/// Maximum amount able to be spent by MediumSpender origin from treasury at once
+		#[pallet::constant]
+		type MaxMediumSpenderSpend: Get<BalanceOf<Self>>;
+		/// Maximum amount able to be spent by BigSpender origin from treasury at once
+		#[pallet::constant]
+		type MaxBigSpenderSpend: Get<BalanceOf<Self>>;
+		/// Maximum amount able to be spent by Treasurer origin from treasury at once
+		#[pallet::constant]
+		type MaxTreasurerSpend: Get<BalanceOf<Self>>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[derive(PartialEq, Eq, Clone, MaxEncodedLen, Encode, Decode, TypeInfo, RuntimeDebug)]
+	#[pallet::origin]
+	pub enum Origin {
+		/// Origin able to dispatch a whitelisted call.
+		WhitelistedCaller,
+		/// Origin able to spend up to MaxTreasurerSpend from the treasury at once.
+		Treasurer,
+		/// Origin able to cancel referenda.
+		ReferendumCanceller,
+		/// Origin able to kill referenda.
+		ReferendumKiller,
+		/// Origin able to spend up to MaxSmallSpenderSpend from the treasury at once.
+		SmallSpender,
+		/// Origin able to spend up to MaxMediumSpenderSpend from the treasury at once.
+		MediumSpender,
+		/// Origin able to spend up to MaxBigSpenderSpend from the treasury at once.
+		BigSpender,
+	}
+
+	impl TryFrom<u8> for Origin {
+		type Error = ();
+		/// TrackId => Origin
+		fn try_from(value: u8) -> Result<Origin, ()> {
+			match value {
+				1 => Ok(Origin::WhitelistedCaller),
+				10 => Ok(Origin::Treasurer),
+				11 => Ok(Origin::ReferendumCanceller),
+				12 => Ok(Origin::ReferendumKiller),
+				13 => Ok(Origin::SmallSpender),
+				14 => Ok(Origin::MediumSpender),
+				15 => Ok(Origin::BigSpender),
+				_ => Err(()),
+			}
+		}
+	}
+
+	impl Into<u16> for Origin {
+		/// Origin => TrackId
+		fn into(self) -> u16 {
+			match self {
+				Origin::WhitelistedCaller => 1,
+				Origin::Treasurer => 10,
+				Origin::ReferendumCanceller => 11,
+				Origin::ReferendumKiller => 12,
+				Origin::SmallSpender => 13,
+				Origin::MediumSpender => 14,
+				Origin::BigSpender => 15,
+			}
+		}
+	}
+
+	#[test]
+	fn origin_track_conversion_is_consistent() {
+		macro_rules! has_consistent_conversions {
+			( $o:expr ) => {
+				let origin_as_u16 = <Origin as Into<u16>>::into($o);
+				let u16_as_u8: u8 = origin_as_u16.try_into().unwrap();
+				let u8_as_origin: Origin = u16_as_u8.try_into().unwrap();
+				assert_eq!($o, u8_as_origin);
+			};
+		}
+		has_consistent_conversions!(Origin::WhitelistedCaller);
+		has_consistent_conversions!(Origin::Treasurer);
+		has_consistent_conversions!(Origin::ReferendumCanceller);
+		has_consistent_conversions!(Origin::ReferendumKiller);
+		has_consistent_conversions!(Origin::SmallSpender);
+		has_consistent_conversions!(Origin::MediumSpender);
+		has_consistent_conversions!(Origin::BigSpender);
+	}
+
+	macro_rules! decl_unit_ensures {
```

Comment:
Some brief documentation about what these macros do would go a long way

Reply:
Oh, it looks like these came out of the kusama runtime. Are we using it correctly, esp. the `runtime-benchmark` part?

Reply:
Yes it was originally taken from [`kusama/governance/origins.rs`](https://github.com/paritytech/polkadot/blob/master/runtime/kusama/src/governance/origins.rs) and right now it lives in [`runtimes/moonbase/governance/origins.rs`](https://github.com/PureStake/moonbeam/blob/master/runtime/moonbase/src/governance/origins.rs).

> Are we using it correctly

I think so, just looked at it again.